### ISSUE TITLE
Can't build `AspNetCoreOData.sln` with .NET 8 SDK

### ISSUE THREAD

**Issue by andrewlock** [OData/AspNetCoreOData#1066]
**Assemblies affected**
Latest on `main`

**Describe the bug**
If you attempt to build with the .NET 8 RC 1 SDK, the build will hang forever

**Reproduce steps**
- Install the .NET 8 SDK RC 1
- `git clone https://github.com/OData/AspNetCoreOData.git`
- `dotnet build .\AspNetCoreOData.sln`
- Builds forever...

**Data Model**
N/A

**EDM (CSDL) Model**
N/A

**Request/Response**
N/A

**Expected behavior**
The build shouldn't hang forever

**Screenshots**
<details><summary>Attempt to build</summary>
<pre><code>dotnet build .\AspNetCoreOData.sln
MSBuild version 17.8.0-preview-23418-03+0125fc9fb for .NET
  Determining projects to restore...
  Restored C:\repos\AspNetCoreOData\sample\ODataSampleCommon\ODataSampleCommon.csproj (in 516 ms).
  Restored C:\repos\AspNetCoreOData\sample\ODataCustomizedSample\ODataCustomizedSample.csproj (in 516 ms).
  Restored C:\repos\AspNetCoreOData\sample\ODataAlternateKeySample\ODataAlternateKeySample.csproj (in 516 ms).
  Restored C:\repos\AspNetCoreOData\sample\ODataDynamicModel\ODataDynamicModel.csproj (in 516 ms).
  Restored C:\repos\AspNetCoreOData\src\Microsoft.AspNetCore.OData\Microsoft.AspNetCore.OData.csproj (in 2.18 sec).
  Restored C:\repos\AspNetCoreOData\sample\ODataRoutingSample\ODataRoutingSample.csproj (in 2.91 sec).
  Restored C:\repos\AspNetCoreOData\test\Microsoft.AspNetCore.OData.Tests\Microsoft.AspNetCore.OData.Tests.csproj (in 5.96 sec)
  .
  Restored C:\repos\AspNetCoreOData\test\Microsoft.AspNetCore.OData.TestCommon\Microsoft.AspNetCore.OData.TestCommon.csproj (in
   5.96 sec).
  Restored C:\repos\AspNetCoreOData\test\Microsoft.AspNetCore.OData.E2E.Tests\Microsoft.AspNetCore.OData.E2E.Tests.csproj (in 6
  .07 sec).
C:\Program Files\dotnet\sdk\8.0.100-rc.1.23463.5\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.target
s(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [C:\repos\AspN
etCoreOData\sample\ODataSampleCommon\ODataSampleCommon.csproj]
C:\Program Files\dotnet\sdk\8.0.100-rc.1.23463.5\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.target
s(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [C:\repos\AspN
etCoreOData\sample\ODataDynamicModel\ODataDynamicModel.csproj]
C:\Program Files\dotnet\sdk\8.0.100-rc.1.23463.5\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.target
s(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [C:\repos\AspN
etCoreOData\sample\ODataAlternateKeySample\ODataAlternateKeySample.csproj]
C:\Program Files\dotnet\sdk\8.0.100-rc.1.23463.5\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.target
s(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [C:\repos\AspN
etCoreOData\sample\ODataRoutingSample\ODataRoutingSample.csproj]
C:\Program Files\dotnet\sdk\8.0.100-rc.1.23463.5\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.target
s(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [C:\repos\AspN
etCoreOData\sample\ODataCustomizedSample\ODataCustomizedSample.csproj]
C:\Program Files\dotnet\sdk\8.0.100-rc.1.23463.5\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.target
s(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [C:\repos\AspN
etCoreOData\test\Microsoft.AspNetCore.OData.E2E.Tests\Microsoft.AspNetCore.OData.E2E.Tests.csproj::TargetFramework=netcoreapp3.
1]
C:\Program Files\dotnet\sdk\8.0.100-rc.1.23463.5\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.target
s(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [C:\repos\AspN
etCoreOData\test\Microsoft.AspNetCore.OData.Tests\Microsoft.AspNetCore.OData.Tests.csproj::TargetFramework=netcoreapp3.1]
C:\Program Files\dotnet\sdk\8.0.100-rc.1.23463.5\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.target
3.1]
C:\Program Files\dotnet\sdk\8.0.100-rc.1.23463.5\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.target
s(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [C:\repos\AspN
etCoreOData\test\Microsoft.AspNetCore.OData.E2E.Tests\Microsoft.AspNetCore.OData.E2E.Tests.csproj::TargetFramework=net6.0]
C:\Program Files\dotnet\sdk\8.0.100-rc.1.23463.5\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.target
etCoreOData\src\Microsoft.AspNetCore.OData\Microsoft.AspNetCore.OData.csproj::TargetFramework=netcoreapp3.1]
s(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [C:\repos\AspN
C:\Program Files\dotnet\sdk\8.0.100-rc.1.23463.5\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.target
s(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [C:\repos\AspN
etCoreOData\test\Microsoft.AspNetCore.OData.TestCommon\Microsoft.AspNetCore.OData.TestCommon.csproj::TargetFramework=net6.0]

... (I cancelled after 10 mins)

Attempting to cancel the build...

Build FAILED.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:03:04.49</code></pre></details>

**Additional context**
This appears to be an issue in the .NET 8 SDK with deeply nested class hierarchies. Just raising here for awareness!
- https://github.com/dotnet/roslyn/issues/70272

**Comment by gathogojr** [comment]
@andrewlock If you're building the solution using Visual Studio 17.7.x version, there's a known issue that we already reported to the responsible team https://developercommunity.visualstudio.com/t/Build-hangs-for-ASPNET-Core-project-whe/10466622.
It's already fix in 17.8.x that's in preview - https://visualstudio.microsoft.com/vs/preview/. You can use the preview to go round the problem pending the GA release.

**Comment by andrewlock** [comment]
Hmmm, I _am_ using the 17.8.x version of Visual Studio 🤔

**Comment by xuzhg** [comment]
Hi, @andrewlock 

Here's my test using the preview version of VS 2022.

<img width="890" alt="image" src="https://github.com/OData/AspNetCoreOData/assets/9426627/d48c395b-cb2f-4092-bb1a-657f39038126">

**Comment by gathogojr** [comment]
@xuzhg Can you try against a .NET8.0 project?

**Comment by xuzhg** [comment]
> @xuzhg Can you try against a .NET8.0 project?

what's .NET 8 prject?

**Comment by gathogojr** [comment]
@xuzhg A project targeting .NET 8.0 framework...

**Comment by gathogojr** [comment]
@xuzhg The issue reporter tried to build a project targeting .NET 8 RC 1 SDK. I don't think we have tested the library against that RC though and whether we should wait for GA

**Comment by julealgon** [comment]
> and whether we should wait for GA

@gathogojr just keep in mind that if you wait for GA and there is a problem with it, you might get a lot more complaints in here from early adopters.

It would be nice if OData was thoroughly tested ASAP with .NET 8 IMHO so that when the time comes, you don't have to delay people's migrations if there are issues.

**Comment by xuzhg** [comment]
by the way, it's duplicated as #1025. @andrewlock Simply change the 'number' mentioned in #1025, and save the changes (let it auto-generate codes), and you can build the source codes no matter which SDK version is. 
@gathogojr I didn't try .NET 8 RC 1 SDK. But my VS2022 preview builder is using '8.0.100-rc.1.23455.8'.
<img width="751" alt="image" src="https://github.com/OData/AspNetCoreOData/assets/9426627/f7f2b097-36c2-4eb1-8b0f-d74733d7ab54">

**Comment by andrewlock** [comment]
> The issue reporter tried to build a project targeting .NET 8 RC 1 SDK. I don't think we have tested the library against that RC though and whether we should wait for GA

@gathogojr I didn't change anything to target .NET 8.0, I just used the .NET 8 RC1 SDK to try to _build_ the AspNetCoreOData.sln in this repository. That should be supported and _isn't_ a bug in the AspNetCoreOData project IMO- it's a regression in the compiler. I'm only raising here as a warning for you, because you _really_ want this to be fixed IMO 😄 

> Simply change the 'number' mentioned in https://github.com/OData/AspNetCoreOData/issues/1025, and save the changes

Thanks @xuzhg, I confirmed that adding `Microsoft.Net.Compilers.Toolset` version 4.6.0 to my sample project also solved the problem. Looks like you have a workaround then 👍 

(FYI, this is my sample project, which is a "minimal" repro for the AspNEtCoreOData issue)

[DeepNestedHierarchy.zip](https://github.com/OData/AspNetCoreOData/files/12856532/DeepNestedHierarchy.zip)
